### DISCUSSION TITLE
issues defining a new board under rp2040 core

### DISCUSSION THREAD

**Discussion by myklemykle** [earlephilhower/arduino-pico#1851] [Q&A]
Hi,  

Apropos of [this thread](https://github.com/earlephilhower/arduino-pico/discussions/1531), I'm trying to develop a standalone board definition for my rp2040-based board.  Arduino docs suggest that this can be files in a completely seperate location (the /hardware subdir of my Arduino sketches directory) and doesn't need to edit files in the main rp2040 distro at all.  

In theory I just need to create a `boards.txt` with custom values, based heavily off of the generic rp2040 entries in the rp2040 distro's boards.txt file . This is the procedure the Arduno docs recommend. However, I'm having some issues getting this to work.  

ATM my problem is that the [platform.txt](https://github.com/earlephilhower/arduino-pico/blob/master/platform.txt) file in the distro uses the Arduino-provided env variable `runtime.platform.path` a lot, to source included headers/libraries/etc . But when I try to compile using my custom boards.txt, `arduino-cli` redefines that variable to point to the location of my custom file, which stands alone in a directory. So all the other included libaries are not found.

I guess I'm wondering why `platform.txt` needs to use `runtime.platform.path` to find the platform files, instead of just hardcoding the paths, since they are distributed together. I would guess that if you're overriding the location of the runtime files you're also overriding platform.txt.  Reading the [Arduino docs](https://arduino.github.io/arduino-cli/0.23/platform-specification/#platform-terminology), they explicitly say `runtime.platform.path` points to the folder where boards.txt lives. (As opposed to the path to the platform files of the runtime, like the name implies.)

Am I pursuing the wrong strategy here?  I have seen Earl's docs on how to merge a new custom board into the rp2040 distro, but is that the only way to install custom board support under arduino-pico?  The "official" Arduino approach seems simpler.

**Comment by earlephilhower** [comment]
Why?  It seems to work OK in both `git` and real IDE installations. :laughing: 

I'm not sure what you mean by "hardcoding the paths" for platform files.  Could you give an example?

For the tools (`runtime.tools.xxx`), this variable allows us to reference the git tree for our binaries and not require installation by the IDE of a toolchain.  They're actually stripped out by the [release scripts](https://github.com/earlephilhower/arduino-pico/blob/master/package/build_boards_manager_package.sh).

**Comment by myklemykle** [comment]
I guess what I'm saying is, the Arduino project's instructions on how to support my custom board -- with just one config file -- are pretty clear, but they seem to be running aground here, and the way `runtime.platform.path` is used in `platform.txt` makes me think it's just not going to work how they describe.

What I mean about the paths is that `platform.txt` uses that variable to set the paths to include dirs, which arduino-cli seems to convert to gcc args. I wondered if there was a better way to get those paths.  But on reflection, I'm not sure how else you'd do it. You can't use paths outside the repo, and I don't see any other better environment variable available. It looks like `runtime.xxx.path` was invented just to support the Teensy distro.

I know that support for custom boards in Arduino has evolved & continues to evolve.  I'm just looking for the supported way to add support for my arduino-pico-based board.  If a PR with mods to a python script is the only supported method, then I'll try that. 

(I just am not sure how to integrate a custom boot2 driver that way.  OTOH I wasn't sure how to integrate it this way either. =/)

**Reply by earlephilhower** [reply]
Ah, got it.  Like you, I'm really not sure how to get the proper paths to GCC without that variable.  You might just want to fork the repo and keep your own private `makeboards.py` to work around the issue at this point. :(

**Reply by myklemykle** [reply]
Well, I sell this board now.  So I'm trying to get working Arduino support that I can distribute to users.  Not 100% necessary, but it would be nice.

**Reply by myklemykle** [reply]
Honestly I don't know why Arduino is changing that variable for custom hardware definitions that refer to other cores. I can't imagine the situation where it would be helpful. OTOH I am willing to bet that changing that behavior would break something for someone.  Maybe I should ask them to define other variables ...something for locating the correct files for the core that's selected in the custom boards.txt .
